vperfetto-sdk.cpp

```diff
@@ -767,22 +825,53 @@ static uint64_t deriveGuestTimeDiffWithGuestAbsoluteTime(
     uint64_t diff = guestBootTimeNs - hostStartTimeNs;
 
     fprintf(stderr, "%s: time diff: %llu\n", __func__, (unsigned long long)diff);
-    return 0;
+    return diff;
 }
 
 static uint64_t deriveGuestTimeDiff(
     const std::vector<char>& guestTrace,
-    const std::vector<char>& hostTrace) {
+    const std::vector<char>& hostTrace,
+    int64_t tscOffset) {
 
     fprintf(stderr, "%s: Deriving guest time diff from guest and host traces\n", __func__);
 
+    // First check for CPU time sync data in both traces.
+    TraceCpuTimeSync hostSync, guestSync;
+    bool hasHostSync = getTraceCpuTimeSync(hostTrace, &hostSync);
+    bool hasGuestSync = getTraceCpuTimeSync(guestTrace, &guestSync);
+    bool sameClock = hasHostSync && hasGuestSync && hostSync.clockId == guestSync.clockId;
+    if (hasHostSync && hasGuestSync && sameClock) {
+        // Transform guest cpuTime to host:
+        guestSync.cpuTime -= tscOffset;
+        fprintf(stderr, "%s: CPU cycles/nanos: host %f, guest %f\n", __func__, hostSync.cpuCyclesPerNano,
+                guestSync.cpuCyclesPerNano);
+        // Guest and host frequency measurement should match.
+        double cyclesPerNano = hostSync.cpuCyclesPerNano;
+        if (hostSync.cpuTime > guestSync.cpuTime) {
+            double cyclesDelta = (double)(hostSync.cpuTime - guestSync.cpuTime);
+            uint64_t offsetNs = (uint64_t)(cyclesDelta / cyclesPerNano);
+            fprintf(stderr, "%s: CPU cycles offset %f seconds\n", __func__, (double)offsetNs / 1000000000.0);
```

Comment:
Should we also log here in each of the cases where host is ahead of or behind the guest? It could make debugging issues much easier.

Reply:
The offset turns out to be pretty large in some cases, like when the host has been up for a long time before the guest starts. If we looked at the monotonic clock snapshot, we could probably detect some kind of mismatch. In practice though this tool is wrapped by a script that makes sure host and guest are traced around the same time.